akka-stream-tests/src/test/scala/akka/stream/scaladsl/WithContextUsageSpec.scala

```diff
@@ -0,0 +1,160 @@
+      val expected = expectedCommits(testData.map(f))
+
+      SourceRefs.createSource(testData)
+        .map(f)
+        .toMat(SinkRefs.createSink(testSink))(Keep.right)
+        .run()
+        .request(expected.size)
+        .expectNextN(expected)
+        .expectComplete()
+    }
+
+    "only commit filtered elements" in {
+      val testData = Vector(Record("k1", "v1"), Record("k2", "v2"))
+      val testSink = TestSink.probe[Committed[Record]]
+      val f: (Record ⇒ Boolean) = record ⇒ record.key == "k2"
+
+      val expected = recordOffsetPairs(testData).filter {
+        case (rec, _) ⇒ f(rec)
+      }.map {
+        case (rec, offset) ⇒ Committed(rec, offset)
+      }
+
+      SourceRefs.createSource(testData)
+        .filter(f)
+        .toMat(SinkRefs.createSink(testSink))(Keep.right)
+        .run()
+        .request(expected.size)
+        .expectNextN(expected)
+        .expectComplete()
+    }
+
+    "only commit when the offset for a batch changes" in {
+      val testData = Vector(Record("k1", "v1"), Record("k2", "v2"), Record("k3", "v3"))
+      val testSink = TestSink.probe[Committed[Record]]
+
+      val f: (Record ⇒ List[Record]) = record ⇒ List(record, record.copy(value = "last"))
+      val expected = Vector(Committed(Record("k1", "last"), 0), Committed(Record("k2", "last"), 1))
+
+      SourceRefs.createSource(testData)
+        .mapConcat(f)
+        .toMat(SinkRefs.createSink(testSink))(Keep.right)
+        .run()
+        .request(testData.size * 2)
+        .expectNextN(expected)
+        .expectComplete()
+    }
+
+    "commit the highest offset when grouped" in {
+      val testData = Vector(Record("k1", "v1"), Record("k2", "v2"), Record("k3", "v3"))
+      val testSink = TestSink.probe[Committed[immutable.Seq[Record]]]
+      val expected = testData
+        .zipWithIndex
+        .grouped(2)
+        .map(_.toSeq)
+        .map(group ⇒ Committed(group.map(_._1), group.last._2))
+        .toVector
+
+      SourceRefs.createSource(testData)
+        .grouped(2)
+        .toMat(SinkRefs.createSink[immutable.Seq[Record]](testSink))(Keep.right)
+        .run()
+        .request(expected.size)
+        .expectNextN(expected)
+        .expectComplete()
+    }
+
+    def expectedCommits[R](expectedRecords: Vector[R]) = expectedRecords.zipWithIndex.map { case (r, index) ⇒ Committed(r, index) }
+    def recordOffsetPairs(testData: Vector[Record]): Vector[(Record, Int)] = testData.zipWithIndex
+  }
+}
+
+case class Offset(value: Int, committable: Consumer.Committable)
+
+object SourceRefs {
+  def createSource(testData: Vector[Record]): SourceWithContext[Offset, Record, NotUsed] =
+    Consumer
+      .committableSource(testData)
+      .withContext(m ⇒ Offset(m.committableOffset.offset, m.committableOffset))
+      .map(_.record)
+
+}
+
+object SinkRefs {
+  def createSink[R](testSink: Sink[Committed[R], Probe[Committed[R]]]): Sink[(R, Op[Offset]), Probe[Committed[R]]] = {
```

Comment:
The idiomatic signature for this functionality would be a `Flow[(R, Op[Offset]), Committed[R], NotUsed]`.

Right now the code seems to only show how to commit offsets (i.e. what is done with the context) but not what is done with the final result element. Would be nice if the example was complete in the sense that something is done with the result elements and only afterwards the offset is committed.

I would recommend to structure it like this:

```
val outputProcessing: Flow[(T, Ctx), Ctx, NotUsed] = ... // does something with the result elements and propagates the context
val commit: Sink[Ctx, Future[Done]] = // does the actual processing (in the test it could return a probe)

theActualFlow
  .via(outputProcessing)
  .toMat(commit)(Keep.right)
  .run()
```

Extra-points if the tests checks that outputProcessing for an element is run before the corresponding offset has been committed.